---
title: "Example code for single-cell analysis with Seurat, day 1"
author: "Kevin Rue-Albrecht"
date: "05/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(tidyverse)
```

# Exercise

## Import scRNA-seq data and create a Seurat object

- Load the `Seurat` package.

```{r}
library(Seurat)
```

- Use the function `Read10X()` to import data in the directory `filtered_feature_bc_matrix/` 
  as an object named `read10x_data`.
  What class of object does the function return?

```{r}
read10x_data <- Read10X("/project/obds/shared/resources/4_r_single_cell/singlecell_seuratday1/filtered_feature_bc_matrix/")

```

```{r}
class(read10x_data)

##The same genes are represented multiple times as they can be found on different chromosomes. you can change it to match EnsembleID by changing gene.column value, but ot may be better to do this later on, once you have found out what is interesting (eg if yu wanted to do a pathway analysis)
```

> Answer:
> 

- Have a look at the object and its structure (e.g., first 15 rows and 6 columns).
  What is a sparse matrix and how does it store data?

```{r}
dim(read10x_data)

read10x_data[1:15, 1:6]

##all the dots stand for zeros because it is a sparse matrix

```

```{r}

```

> Answer:
> 

- How many features and barcodes (i.e., cells) are present in the data set?

```{r}
dim(read10x_data)

```

> Answer: there are 33538 features (genes) and 5155 barcodes (most of which are probably cells)
> 

- Create a Seurat object using the function `CreateSeuratObject()`
  and the object `read10x_data`.
  Name the object `seurat_object`.
  Include features detected in at least 3 cells,
  and cells where at least 200 features detected.
  Name the project `pbmc5k`.
  How many features and barcodes are left in the Seurat object?

```{r}
##As far as seurat is concerned a project is one dataset. It is only really useful if you need to merge datasets. The project argument gives your cells an identifier so that if you are merging datasets you know where it came from

?CreateSeuratObject
seurat_object <- CreateSeuratObject(
  counts = read10x_data, 
  project = "pbmc5k", 
  min.cells = 3, 
  min.features = 200)

seurat_object
```

> Answer:
> 

- How many features and cells were filtered out when you created the Seurat object?

```{r}
dim(read10x_data) - dim(seurat_object)
```

> 

# Exercise

## Accessing the contents of a Seurat object

- Query the name of the default assay in the Seurat object.

```{r}

DefaultAssay(seurat_object)
##or can use
seurat_object@active.assay

```

- List the names of assays available in the Seurat object.

```{r}

Assays(seurat_object)
```

- Display the first six rows and six columns of the RNA assay data.
  What function do you use?
  Which arguments do you need to supply, and which ones are optional?

```{r}
?GetAssayData
GetAssayData(seurat_object[["RNA"]], slot = "data") [1:6, 1:6]
GetAssayData(seurat_object, assay = "RNA", slot = "data")[1:6, 1:6]

seurat_object@assays$RNA@data[1:6, 1:6]

```

> Answer:
> 

- Display the entire data.frame of per-cell metadata (first six rows).
  What column names do you see?

```{r}
head(seurat_object[[]])




```

> Answer:
> 

- Fetch one column of metadata using `[[`.
  What type of object do you get back?

```{r}
seurat_object[["nCount_RNA"]]

class(seurat_object[["nCount_RNA"]])
```

> Answer:
> 

- Instead,fetch the same column of metadata using `$`.
  What type of object do you get back this time?

```{r}

seurat_object$nCount_RNA
class(seurat_object$nCount_RNA)
```

> Answer:
> 

- Use the function `FetchData()` to access the library size and expression of the feature named `"LYZ"` (first six rows).
  What type of object do you get back?

```{r}
?FetchData
FetchData(seurat_object, vars = c("LYZ", "nCount_RNA"))[1:6,]
```

> 

# Demo

## Common operations on Seurat objects

`WhichCells()` returns the names of cells that match a logical expression.

```{r}
WhichCells(seurat_object, expression = LYZ > 500)
```

`VariableFeatures()` returns the names of variable features (for a given assay, if computed).

```{r}
VariableFeatures(seurat_object)
```

`subset()` returns a new Seurat object restricted to certain features and cells.

```{r}
subset(
    x = seurat_object,
    cells = WhichCells(seurat_object, expression = LYZ > 500),
    features = VariableFeatures(object = seurat_object)
)
```

# Exercise

## Quality control and visualisation

- The library size and number of features detected per cell is already present in the Seurat object.
  Use the function `VlnPlot()` to display them in a single violin plot.

```{r}
?VlnPlot
VlnPlot(seurat_object, features = c("nCount_RNA", "nFeature_RNA"))
```

- Use the function `PercentageFeatureSet()` to compute the fraction of reads
  assigned to mitochondrial genes in each cell.
  Store the metric in the cell metadata of the Seurat object, under the name `"percent_mt"`.
  Visualise this new metric alongside the previous two in a new violin plot.

```{r}

##here when we are doing the [[]] in the assigning the name, this is adding a new column to the object with the name we are giving inside the square brackets
seurat_object[["percent.mt"]] <- PercentageFeatureSet(seurat_object, pattern = "^MT")
VlnPlot(seurat_object, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"))
```

- Visualise a scatter plot of the proportion of mitochondrial UMIs against the library size in each cell.

```{r}

?FeatureScatter
FeatureScatter(seurat_object, feature1 = "percent.mt", feature2 = "nCount_RNA")
```

- Create a new Seurat object, called `seurat_after_qc`, that is subsetted to cells that have more than 4,500 UMI counts, less than 15% of UMI counts assigned to mitochondrial features, and more than 1,500 features detected.
  How many cells were removed in this step?

```{r}
?subset
seurat_after_qc <- subset(seurat_object,
                          subset = nCount_RNA > 4500 & percent.mt <15 & nFeature_RNA >1500
    )
seurat_after_qc
```

```{r}
dim(seurat_object) - dim(seurat_after_qc)
```

> Answer:
> 

# Exercise

## Normalisation

- Normalise the RNA assay of the Seurat object (after quality control) using the `"LogNormalize"` method.

```{r}
?NormalizeData
seurat_after_qc <- NormalizeData(seurat_after_qc, normalization.method = "LogNormalize")
head(seurat_after_qc)
```

## Bonus

- Visualise the distribution of raw counts and normalised data for a feature of your choice.

```{r}

GetAssayData(seurat_after_qc, slot = "counts")
?FetchData
lyznorm <- FetchData(seurat_after_qc, vars = "LYZ", slot = "data")
lyzraw <- FetchData(seurat_after_qc, vars = "LYZ", slot = "counts")

lyzraw
ggplot_lyz_raw <- ggplot(lyzraw, aes(x = LYZ)) +
    geom_histogram() +
    coord_cartesian(ylim = c(0, 500)) +
    cowplot::theme_cowplot()

ggplot_lyz_raw

ggplot_lyz_normalised <- ggplot(lyznorm, aes(x = LYZ)) +
    geom_histogram() +
    coord_cartesian(ylim = c(0, 500)) +
    cowplot::theme_cowplot()

ggplot_lyz_normalised
cowplot::plot_grid(ggplot_lyz_raw, ggplot_lyz_normalised, ncol = 1)
```

# Exercise

## Variable features and scaling

- Identify variable features in the normalised RNA assay of the Seurat object.
  Use the `"vst"` method and select the 2,000 most variable features.
  What does this subsetting do, and what are our motivations for doing it?

```{r}
?FindVariableFeatures
seurat_after_qc <- FindVariableFeatures(seurat_after_qc, selection.method = "vst", nfeatures = 2000)
```

> Answer:
> 

- What is the function to display the name of variable features in a Seurat object (e.g., first 10)?
  How can you control which assay the variable features are pull from?

```{r}
?VariableFeatures()

VariableFeatures(seurat_after_qc)[1:10]

```

> Answer:
> 

- Use the function `VariableFeaturePlot()` to visualise the scatter plot of standardised variance against average expression.
  How would you use this plot?

```{r}
?VariableFeaturePlot
VariableFeaturePlot(seurat_after_qc)
```

> Answer:
> 

- Scale the normalised RNA assay of the Seurat object, regressing the library size and the fraction of UMI counts assigned to mitochondrial features.
  What are the motivations for removing those two sources of variation?

```{r}
?ScaleData

seurat_after_qc <- ScaleData(seurat_after_qc, vars.to.regress = c("nCount_RNA", "percent.mt"))
##regressing out library size and mitochondrial gene fraction to account for the differences/variations caused by these variables
```

> Answer:
> 

# Exercise

## Dimensionality reduction

- Run a principal component analysis on the Seurat object.
  Which features are used by the method in the default settings?
  How could you change this?
  How do you read the message output of the function `RunPCA()`?

```{r}
?RunPCA
seurat_after_qc <- RunPCA(seurat_after_qc)
```

> Answer:
> 

- List the names of dimensionality reduction results available in the Seurat object.

```{r}
Reductions(seurat_after_qc)

```

- Use `PCAPlot()` or `DimPlot()` to produce a scatterplot of the first and second PCA components.

```{r}
?PCAPlot

PCAPlot(seurat_after_qc)
```

## Bonus

- Make a scatterplot of the first and second PCA components yourself using `ggplot2`.

```{r}
# Use this code chunk to prepare a data.frame for ggplot2
?FetchData
pca_data <- FetchData(seurat_after_qc, vars = c("PC_1", "PC_2"))
head(pca_data)
```

```{r}
ggplot(pca_data, aes(x = PC_1, y = PC_2)) +
    geom_point(size = 0.2) +
    cowplot::theme_cowplot()
```

- Visualise the amount of variance explained the top principal components (number of your choice).
  How many principal components would you use for downstream analyses?

```{r}
?ElbowPlot
ElbowPlot(seurat_after_qc, ndims = 50) + coord_cartesian(ylim = c(0, 15)) ##this bit allows you to control the limits of the y axis
```

> 

- Run the UMAP technique on your selected number of principal components and visualise the result as a scatterplot.

```{r}
?RunUMAP
seurat_after_qc <- RunUMAP(seurat_after_qc, dims = 1:18, reduction = "pca")
UMAPPlot(seurat_after_qc)

?saveRDS
saveRDS(seurat_after_qc, file = "SeuratAfterQC_Feb2023.rds" )
```

# Exercise

## Clustering

- Compute the graph of nearest neighbours using the function `FindNeighbors()`.
  Which principal components are used by default?
  Instead, specify the number of principal components that you have chosen earlier.

```{r}

?FindNeighbors
##The command automatically uses dimensions 1:10 so if we want it to use a different number we need to specify 
 seurat_after_qc <- FindNeighbors(seurat_after_qc, reduction = "pca", dims = 1:18, k.param = 20)
```

> Answer:
> 
> The help page states that the function `FindNeighbors()` uses principal components 1 through 10, by default.

- What are the names of the nearest neighbour graphs that are now stored in the Seurat object?

```{r}
Graphs(seurat_after_qc)

##nn = nearest neighbour, snn = shared nearest neighbour
```

- Finally, compute cluster labels.
  What is the default setting for the `resolution` argument?
  Instead, set it to `0.5`.
  Do you expect more or fewer clusters following that change?
  What other parameters would you also try to experiment with?

```{r}

?FindClusters
##the default resolution is 0.8, so here we have changed it to 0.5
##You can also try multiple resolutions 
seurat_after_qc <- FindClusters(seurat_after_qc, graph.name = "RNA_snn", resolution = 0.5, algorithm = 1)
```

> 

- Visualise the cluster labels on the UMAP scatter plot.
  How would you describe the agreement between the UMAP layout and the clustering results?

```{r}
?DimPlot

UMAPPlot(seurat_after_qc, group.by = "RNA_snn_res.0.5", label = TRUE)
```

> 

# Exercise

## Identify cluster markers

- Use the function `FindAllMarkers()` to identify
  positive markers for all clusters,
  filtering markers that are detected in at least 25% of the cluster,
  and with a log fold-change greater than `0.25`.
  Assign the result to an object named `seurat_markers_all`.
  What is the class of that object?
  How do you control the set of clusters that are used?

```{r}
?FindAllMarkers

seurat_markers_all <- FindAllMarkers(seurat_after_qc, logfc.threshold = 0.25, min.pct = 0.25)
    
    
    
    

class(seurat_markers_all)
```

> Answer:
> 

- How do you read the contents of the object `seurat_markers_all`?
  How do you know which features are the markers of each cluster?

```{r}
head(seurat_markers_all)
##pct.1 is telling you that 85% of the cells in the cluster are expressing the gene, pct.2 is what percentage of genes outside of the cluster are expressing the gene
```


> Answer:
> 

- Filter and display the top 10 markers for cluster 3.

```{r}

seurat_markers_all %>%
  filter(cluster == 3) %>%
  arrange(desc(abs(avg_log2FC))) %>%
  slice_head(n = 10) -> markersCluster3 ##this is how you can do assignment at the end of your pipe
markersCluster3

seurat_markers_all %>%
  filter(cluster == 3) %>%
  arrange(desc((avg_log2FC))) %>% ##here we got rid of the abs because we want the positively expressed genes 
  slice_head(n = 4) %>% 
  pull(gene) -> genesCluster3 

##could also use slice_max
```

- Visualise the expression of the top 4 marker for cluster 3 on a UMAP layout.

```{r}
?FeaturePlot


FeaturePlot(seurat_after_qc, label = TRUE, features = genesCluster3)

```

- Visualise the expression of those same 4 marker as a violin plot.
  Do you have any particular preference between the two types of plots?

```{r}
VlnPlot(seurat_after_qc, features = genesCluster3)

```

> Answer:
> 

- Use `DoHeatmap()` to visualise the top 10 (positive) markers for each cluster.
  Hint: you may want to use the function `dplyr::group_by()`.

```{r, fig.width=15, fig.height = 10}
markers_top10_clusters <- seurat_markers_all %>%
  group_by(cluster) %>%
  arrange(desc((avg_log2FC))) %>%
  slice_head(n = 10)
  
markers_top10_clusters


DoHeatmap(seurat_after_qc, features = markers_top10_clusters$gene)
```
